traffic_ops/traffic_ops_golang/iso/iso.go

```diff
@@ -0,0 +1,368 @@
+	MgmtIPGateway net.IP          `json:"mgmtIpGateway"`
+	MgmtInterface string          `json:"mgmtInterface"`
+	Stream        boolStr         `json:"stream"`
+}
+
+func (i *isoRequest) fqdn() string {
+	fqdn := i.HostName
+	if i.DomainName != "" {
+		fqdn += "." + i.DomainName
+	}
+	return fqdn
+}
+
+// validate returns an empty slice if the isoRequest is valid. Otherwise,
+// it returns a slice of error messages.
+func (i *isoRequest) validate() []string {
+	var errs []string
+	addErr := func(msg string) { errs = append(errs, msg) }
+
+	if i.OSVersionDir == "" {
+		addErr("osversionDir is required")
+	}
+	if i.HostName == "" {
+		addErr("hostName is required")
+	}
+	if i.DomainName == "" {
+		addErr("domainName is required")
+	}
+	if i.RootPass == "" {
+		addErr("rootPass is required")
+	}
+	if !i.DHCP.isSet {
+		addErr("dhcp is required")
+	}
+	if i.InterfaceMTU == 0 {
+		addErr("interfaceMtu is required")
+	}
+	if i.Disk == "" {
+		addErr("disk is required")
+	}
+	if len(i.MgmtIPAddress) > 0 {
+		if i.MgmtInterface == "" {
+			addErr("mgmtInterface is required when mgmtIpAddress is provided")
+		}
+		if len(i.MgmtIPGateway) == 0 {
+			addErr("mgmtIpGateway is required when mgmtIpAddress is provided")
+		}
+	}
+	if v, ok := i.DHCP.val(); ok && !v {
+		if len(i.IPAddr) == 0 {
+			addErr("ipAddress is required if DHCP is no")
+		}
+		if len(i.IPNetmask) == 0 {
+			addErr("ipNetmask is required if DHCP is no")
+		}
+		if len(i.IPGateway) == 0 {
+			addErr("ipGateway is required if DHCP is no")
+		}
+	}
+
+	// Use of stream is deprecated it will always stream
+	// regardless if this is set or not.
+	if v, ok := i.Stream.val(); ok && !v {
+		addErr("stream request parameter is deprecated")
+	}
+
+	return errs
+}
+
+// validateOSDir ensures that the OSDir value corresponds to a
+// valid directory, as determined by the osversions.json config
+// file. Since this directory is acted upon during ISO generation,
+// it's important that it be valid.
+func (i *isoRequest) validateOSDir(tx *sqlx.Tx) (bool, error) {
+	validOSVersions, err := getOSVersions(tx)
+	if err != nil {
+		return false, fmt.Errorf("unable to read osversions configuration: %v", err)
+	}
+
+	var isValid bool
+	for _, validOSDir := range validOSVersions {
+		if i.OSVersionDir == validOSDir {
+			isValid = true
+			break
+		}
+	}
+
+	return isValid, nil
+}
+
+// boolStr is used to decode boolean strings (e.g. "yes") as
+// part of a JSON response. Part of the /isos JSON request
+// generated by TrafficPortal uses this format.
+// If an unrecognize or empty string is given, then
+// the 'val' and 'isSet' fields will be false. Otherwise,
+// 'isSet' will be true.
+type boolStr struct {
+	isSet bool // false if UnmarshalText is given an unrecognized value
+	v     bool
+}
+
+// UnmarshalText decodes strings representing boolean values.
+// It nevers returns an error to allow for all validation errors
+// to be grouped together.
+func (b *boolStr) UnmarshalText(text []byte) error {
```

Comment:
Why doesn't this just return an error in the case that the text doesn't represent a string boolean?